### PR TITLE
Introduce DI & migrate tools builders

### PR DESCRIPTION
**Pull request by gohabereg**

### CODE REVIEW

File: packages/core/src/tools/builders/BaseToolBuilder.ts
Comment (neSpecc) [inline packages/core/src/tools/builders/BaseToolBuilder.ts:1]: lets call them "facade"

File: packages/core/src/tools/ToolsManager.ts [hunk 1]
```diff
@@ -1,27 +1,211 @@
...
+              config: config,
+            });
+
+            const tool = this.#factory.get(toolName);
+
+            if (tool.isInline()) {
+              /**
+               * Some Tools validation
+               */
+              const inlineToolRequiredMethods = ['render'];
```
Comment (e11sy) [inline packages/core/src/tools/ToolsManager.ts:108]: maybe move this constant to the sdk package?

File: packages/core/src/index.ts [hunk 2]
```diff
@@ -21,6 +23,8 @@ const DEFAULT_HOLDER_ID = 'editorjs';
  * - adds Blocks accodring to model updates
  */
 export default class Core {
+  #id = Math.floor(Math.random() * 1e10).toString();
```
Comment (neSpecc) [inline packages/core/src/index.ts:26]: you can use local variable instead

File: packages/core/src/tools/ToolsManager.ts [hunk 3]
```diff
@@ -1,27 +1,211 @@
...
+
+                this.#unavailableTools.set(tool.name, tool);
+
+                return;
+              }
+            }
+
+            this.#availableTools.set(toolName, tool);
+
+            console.log(this.#availableTools.entries());
```
Comment (e11sy) [inline packages/core/src/tools/ToolsManager.ts:129]: seems redundant

File: packages/core/src/tools/builders/BlockToolBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+import { type InlineToolBuilder } from './InlineToolBuilder.js';
+import { ToolType } from './ToolType.js';
+import { type BlockTuneBuilder } from './BlockTuneBuilder.js';
+import { ToolsCollection } from './ToolsCollection.js';
+import { BlockToolConstructor as BlockToolConstructable, BlockTool as IBlockTool } from '../../entities/BlockTool.js';
+
+/**
+ * Class to work with Block tools constructables
+ */
+export class BlockToolBuilder extends BaseToolBuilder<ToolType.Block, IBlockTool> {
```
Comment (neSpecc) [inline packages/core/src/tools/builders/BlockToolBuilder.ts:20]: we have `api.tools.getBlockTools()` method that exposes builders (facades), so they probably should have interfaces in `sdk`

File: packages/core/src/tools/ToolsManager.ts (hunk 1 above)
Comment (gohabereg) [reply packages/core/src/tools/ToolsManager.ts:108]: We can do that later

File: packages/core/src/tools/facades/BlockToolFacade.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+  protected declare constructable: BlockToolConstructable;
+
+  /**
+   * Creates new Tool instance
+   * @param options
+   * @param options.data - Tool data
+   * @param options.block - BlockAPI for current Block
+   * @param options.readOnly - True if Editor is in read-only mode
+   */
+  public create({ data, block, readOnly, adapter }: { data: BlockToolData, block: BlockAPI, readOnly: boolean, adapter: BlockToolAdapter }): IBlockTool {
```
Comment (neSpecc) [inline packages/core/src/tools/facades/BlockToolFacade.ts:48]:
```suggestion
  public create({ data, block, readOnly, adapter }: Pick<BlockToolConstructorOptions, 'adapter' | 'data' | 'block' | 'readOnly'>): IBlockTool {
```

File: packages/core/src/tools/facades/BlockToolFacade.ts [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+  /**
+   * Returns true if read-only mode is supported by Tool
+   */
+  public get isReadOnlySupported(): boolean {
+    return this.constructable[InternalBlockToolSettings.IsReadOnlySupported] === true;
+  }
+
+  /**
+   * Returns true if Tool supports linebreaks
+   * @todo check if we still need this as BlockToolConstructable doesn't have line breaks option
```
Comment (neSpecc) [inline packages/core/src/tools/facades/BlockToolFacade.ts:68]: Code tool [uses](https://github.com/editor-js/code/blob/master/src/index.ts#L89) it.

File: packages/core/src/tools/facades/ToolsFactory.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  InlineToolConstructable,
+  BlockTuneConstructable
+} from '@editorjs/editorjs';
+
+type ToolConstructor = typeof InlineToolFacade | typeof BlockToolFacade | typeof BlockTuneFacade;
+
+export type UnifiedToolConfig = Record<string, Omit<ToolSettings, 'class'> & {
+  class: ToolConstructable;
+  isInternal?: boolean;
+}>;
```
Comment (neSpecc) [inline packages/core/src/tools/facades/ToolsFactory.ts:19]: move to `/entities/`

File: packages/core/src/tools/facades/ToolsFactory.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+export class ToolsFactory {
+  /**
+   * Tools configuration specified by user
+   */
+  private config: UnifiedToolConfig;
+
+  /**
+   * EditorJS API Module
+   */
+  private api: any;
```
Comment (neSpecc) [inline packages/core/src/tools/facades/ToolsFactory.ts:33]: you can 
```ts
import { API } from '@editorjs/editorjs'
```

File: packages/core/src/tools/facades/BlockToolFacade.ts (hunk 6 above)
Comment (gohabereg) [reply packages/core/src/tools/facades/BlockToolFacade.ts:70]: We don't have it in the types, so we'll need to update types first

File: packages/core/src/tools/facades/BlockToolFacade.ts (hunk 5 above)
Comment (gohabereg) [reply packages/core/src/tools/facades/BlockToolFacade.ts]: image

File: packages/core/src/tools/facades/ToolsFactory.ts (hunk 8 above)
Comment (gohabereg) [reply packages/core/src/tools/facades/ToolsFactory.ts:43]: Not in this PR, when API will be implemented in the core package

File: packages/core/src/entities/BlockTool.ts
Comment (neSpecc) [inline packages/core/src/entities/BlockTool.ts:1]: what is this

File: packages/core/src/entities/UnifiedToolConfig.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import type { ToolSettings, ToolConstructable } from '@editorjs/editorjs';
+import type { BlockToolConstructor, InlineToolConstructor } from '@editorjs/sdk';
+
+export type UnifiedToolConfig = Record<string, Omit<ToolSettings, 'class'> & {
```
Comment (neSpecc) [inline packages/core/src/entities/UnifiedToolConfig.ts:4]:
```suggestion
/**
 * Users can pass tool's config in two ways:
 *   toolName: ToolClass
 *   or
 *   toolName: { 
 *     class: ToolClass,
 *     // .. other options      
 *   }   
 *
 * This interface unifies these variants to a single format
 */
export type UnifiedToolConfig = Record<string, Omit<ToolSettings, 'class'> & {
```

File: packages/core/src/entities/UnifiedToolConfig.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+  /**
+   * Tool constructor
+   */
+  class: ToolConstructable | BlockToolConstructor | InlineToolConstructor;
+
+  /**
+   * Specifies if tool is internal
```
Comment (neSpecc) [inline packages/core/src/entities/UnifiedToolConfig.ts:11]: describe when it can be omitted

File: packages/core/src/tools/ToolsManager.ts [hunk 11]
```diff
@@ -1,42 +1,223 @@
...
+   * @param editorConfig.tools - Tools configuration passed by user
+   */
+  constructor(@Inject('EditorConfig') editorConfig: EditorConfig) {
+    this.#config = this.#prepareConfig(editorConfig.tools ?? {});
+
+    this.#validateTools();
+
+    this.#factory = new ToolsFactory(this.#config, editorConfig, {});
+
+    void this.#prepareTools();
```
Comment (neSpecc) [inline packages/core/src/tools/ToolsManager.ts:87]: im not sure that this promise should be ignored. Maybe you can await it in somewhere Core

File: packages/core/src/tools/ToolsManager.ts [hunk 12]
```diff
@@ -1,42 +1,223 @@
...
+   * @returns Promise<void>
+   */
+  #prepareTools(): Promise<void> {
+    const promiseQueue = new PromiseQueue();
+
+    Object.entries(this.#config).forEach(([toolName, config]) => {
+      if (isFunction(config.class.prepare)) {
+        void promiseQueue.add(async () => {
+          try {
+            await config.class.prepare!({
```
Comment (neSpecc) [inline packages/core/src/tools/ToolsManager.ts:101]: The `!` here looks unexpected

File: packages/core/src/tools/ToolsManager.ts (hunk 11 above)
Comment (gohabereg) [reply packages/core/src/tools/ToolsManager.ts]: I made the method public, let's await it once `isReady` api is implemented

File: packages/core/src/tools/ToolsManager.ts (hunk 12 above)
Comment (gohabereg) [reply packages/core/src/tools/ToolsManager.ts:119]: Added a comment

**Review by e11sy — APPROVED** [review]

**Review by neSpecc — APPROVED** [review]

**Pull request merged**